File: src/lib/server/services/kitchen.ts
import { menuItemsById } from "./menu-items";
import { currentBusinessDate, parseOrderRequest, readSalesRows } from "./sales";
import type { KitchenBoard, KitchenTicket } from "#lib/types/api";

const maxTickets = 20;

export function getKitchenBoard(): KitchenBoard {
  const rows = readSalesRows(currentBusinessDate()).slice(0, maxTickets);

  const tickets: KitchenTicket[] = rows.map((row) => {
    const request = parseOrderRequest(row.request_json);
    return {
      orderNumber: row.order_number,
      createdAt: row.created_at,
      notes: request.notes,
      lines: request.items
        .filter((line) => menuItemsById.get(line.menuItemId)?.printGroup !== "customer")
        .map((line) => ({
          menuItemId: line.menuItemId,
          name: menuItemsById.get(line.menuItemId)?.name ?? line.menuItemId,
          quantity: line.quantity,
          notes: line.notes
        }))
    };
  });

  return {
    serverTime: new Date().toISOString(),
    tickets: tickets.filter((ticket) => ticket.lines.length > 0)
  };
}

File: internal/pos/kitchen.go
package pos

// KitchenTicketLine is one item on a kitchen ticket.
type KitchenTicketLine struct {
	MenuItemID string `json:"menuItemId"`
	Name       string `json:"name"`
	Quantity   int    `json:"quantity"`
	Notes      string `json:"notes,omitempty"`
}

// KitchenTicket is one order on the kitchen display.
type KitchenTicket struct {
	OrderNumber int                 `json:"orderNumber"`
	CreatedAt   string              `json:"createdAt"`
	Notes       string              `json:"notes,omitempty"`
	Lines       []KitchenTicketLine `json:"lines"`
}

// KitchenBoard is the body of GET /api/kitchen.
type KitchenBoard struct {
	ServerTime string          `json:"serverTime"`
	Tickets    []KitchenTicket `json:"tickets"`
}

const maxKitchenTickets = 20

// GetKitchenBoard builds the kitchen display of today. It keeps the newest
// twenty orders and drops the items that only the customer receipt shows.
func (service *OrderService) GetKitchenBoard() (KitchenBoard, error) {
	rows, err := ReadSalesRows(service.DB, CurrentBusinessDate(service.Now))
	if err != nil {
		return KitchenBoard{}, err
	}
	if len(rows) > maxKitchenTickets {
		rows = rows[:maxKitchenTickets]
	}

	tickets := make([]KitchenTicket, 0, len(rows))
	for _, row := range rows {
		request := parseStoredRequest(row.RequestJSON)
		lines := make([]KitchenTicketLine, 0, len(request.Items))

		for _, line := range request.Items {
			if item, found := MenuItemByID(line.MenuItemID); found && item.PrintGroup == PrintGroupCustomer {
				continue
			}
			lines = append(lines, KitchenTicketLine{
				MenuItemID: line.MenuItemID,
				Name:       MenuItemName(line.MenuItemID),
				Quantity:   line.Quantity,
				Notes:      line.Notes,
			})
		}

		if len(lines) == 0 {
			continue
		}
		tickets = append(tickets, KitchenTicket{
			OrderNumber: row.OrderNumber,
			CreatedAt:   row.CreatedAt,
			Notes:       request.Notes,
			Lines:       lines,
		})
	}

	return KitchenBoard{
		ServerTime: service.Now().UTC().Format(timestampLayout),
		Tickets:    tickets,
	}, nil
}
